In [ ]:
# !pip3 install pyarrow

In [ ]:
# !pip3 install fastparquet

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 684.1/684.1 kB 6.9 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.7/1.7 MB 9.9 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 2/2 [fastparquet]


터미널에 git 연결하고 이름이랑 이메일 입력하고 개인 이름 브랜치 연결해서 커밋
- git init
- git config --global user.name {깃허브 아이디}
- git config --global user.email {깃허브 email}
- git checkout {자기 브랜치 이름{

In [1]:
import pandas as pd

```
### 데이터 #2 (유저 이벤트 데이터**)**

- 고급 프로젝트 → dump-**hackle**-202411211621.sql
- 이벤트 데이터 대상 기간: 2023-07-11 ~ 2023-08-04 **(약 4주)**
    - 원본 이벤트 데이터의 양이 300GB 정도로 과도하게 많아 일부 데이터만 제공합니다.
    - 전체 기간 중 데이터의 용량이나 관리하는 이벤트, 데이터 형식이 제각각 달라 데이터 형태가 가장 비슷하고, 정보량(데이터 크기)이 많은 기간만 선정하여 제공하였습니다.
```

In [27]:
device_properties_df = pd.read_parquet('hackle/device_properties.parquet')  # 엔진은 자동으로 선택됨
device_properties_df.head()

id                             device_id device_model device_vendor
0   1  000007C9-E103-4EB5-9777-A9084D4952DF   iPhone14,7         Apple
1   2  00002245-458F-4CDD-8533-B448CD43DBD2   iPhone14,7         Apple
2   3  00012620-313A-4502-9F8D-8DAB7443215B   iPhone14,5         Apple
3   4  000137bc-80de-4bb5-b61d-df7f217a4501     SM-F711N       samsung
4   5  000227D6-B782-4367-91C4-486B76DF9E37   iPhone12,3         Apple

In [29]:
device_properties_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 252380 entries, 0 to 252379
Data columns (total 4 columns):
 #   Column         Non-Null Count   Dtype 
---  ------         --------------   ----- 
 0   id             252380 non-null  int64 
 1   device_id      252380 non-null  object
 2   device_model   252380 non-null  object
 3   device_vendor  252380 non-null  object
dtypes: int64(1), object(3)
memory usage: 7.7+ MB


In [30]:
device_properties_df.describe(include='O')

device_id device_model device_vendor
count                                 252380       252380        252380
unique                                251720          522            36
top     5dab2e2f-54fd-4c4d-9996-1cccfa6c7428   iPhone14,5         Apple
freq                                       2        27656        174444

In [31]:
device_properties_df['device_vendor'].unique()

array(['Apple', 'samsung', 'Lenovo', 'LGE', 'Xiaomi', 'blackshark',
       'Google', 'FYD', 'Samsung', 'Chinoe', 'vivo', 'ZTE', 'OnePlus',
       'motorola', 'Foxconn', 'Shanling', 'Teclast', 'Nothing', 'Amazon',
       'LENOVO', 'HUAWEI', 'vsmart', 'Digital', 'Sony', 'bnd',
       'Alldocube', 'Teclast_korea', 'incar', 'asus', 'E-dong', 'realme',
       'APEX', 'Razer', 'HXY', 'IMUZ', 'iMUZ'], dtype=object)

In [33]:
device_properties_df[device_properties_df['device_id'].duplicated()]

id                             device_id device_model  \
17          18  0006bed8-512b-48f9-9b33-f79bb1e225a8     SM-S911N   
705        706  00bf3a4e-4ef9-4a6a-8558-cc5f68d4041d     SM-A716S   
1719      1720  01BBBADC-073C-4AEC-8C00-52F0A29577B4   iPhone14,4   
1825      1826  01D2DB25-5EF6-447C-B3FA-63156C4DCBCF   iPhone14,5   
2135      2136  021fd2a9-a0e5-42e8-b9d1-0c2244101869     SM-S918N   
...        ...                                   ...          ...   
250452  250453  FDFA56F4-2289-4542-AC99-4E68FA63D91A   iPhone14,4   
250542  250543  fe135415-8a57-44da-91e6-2b2ac0c4ee61     SM-N950N   
250799  250800  fe599a47-85ad-467b-b3e5-ec557117ec80     SM-G991N   
251275  251276  FED7434C-C9F0-4FF2-BB80-8520E37A1E01   iPhone13,1   
252294  252295  ffeae563-c45e-4f68-85e0-f0373dd940cf     SM-S918N   

       device_vendor  
17           samsung  
705          samsung  
1719           Apple  
1825           Apple  
2135         samsung  
...              ...  
250452         Apple  
250542       samsung  
250799       samsung  
251275         Apple  
252294       samsung  

[660 rows x 4 columns]

- device_id 의 중복은 최대가 2개이며 660개의 데이터. 
- 아이폰 14의 유저, 아이폰 기종이 가장 많은 비율을 차지
- nan 결측치 없음

---

In [ ]:
# 약 4주간의 헥클 이벤트 발생 목록
hackle_events_df = pd.read_parquet('hackle/hackle_events.parquet') 
hackle_events_df.head()

event_id      event_datetime  \
0  00000533-3f1c-4b3b-81f1-0c8f35754b4e 2023-07-18 19:40:17   
1  00000716-27e9-4e72-a602-d0ce61784b06 2023-07-18 21:07:24   
2  000007c8-68ce-40e6-9b1e-f0e34e8ff9cc 2023-08-06 20:18:03   
3  00000981-5e2a-4111-993e-4f1891ad9a53 2023-08-05 01:46:10   
4  00000a7a-ba72-4332-b4a9-7910670aaeb2 2023-07-24 15:03:37   

                         event_key                            session_id  \
0                   $session_start          4OzYh3seq3VKytpSn5pvQkZNQii1   
1              click_question_open          8QXy31PQxbW9qLzq0Y1dhR8Ypm52   
2  click_bottom_navigation_profile  6bcea65d-9f40-46fc-888c-700fe707483f   
3                        view_shop          XVYNT6zfhFWqIg9omwg2AHDjTLx2   
4      click_bottom_navigation_lab          XFB2SPiGfjbVhvJ3Q3DBsaT3m2B3   

                                     id item_name page_name  friend_count  \
0  00000533-3f1c-4b3b-81f1-0c8f35754b4e                               NaN   
1  00000716-27e9-4e72-a602-d0ce61784b06                              64.0   
2  000007c8-68ce-40e6-9b1e-f0e34e8ff9cc                              26.0   
3  00000981-5e2a-4111-993e-4f1891ad9a53                              61.0   
4  00000a7a-ba72-4332-b4a9-7910670aaeb2                             119.0   

   votes_count  heart_balance  question_id  
0          NaN            NaN          NaN  
1        436.0         4830.0          NaN  
2        174.0         4729.0          NaN  
3         44.0          142.0          NaN  
4        545.0         3287.0          NaN

In [37]:
# 최대 컬럼 정보 제한을 늘리기
pd.set_option('display.max_info_columns', 100)
pd.set_option('display.max_columns', None)  # 모든 컬럼 보이기
pd.set_option('display.expand_frame_repr', False)

print(hackle_events_df.info(verbose=True))

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 11441319 entries, 0 to 11441318
Data columns (total 11 columns):
 #   Column          Dtype         
---  ------          -----         
 0   event_id        object        
 1   event_datetime  datetime64[ms]
 2   event_key       object        
 3   session_id      object        
 4   id              object        
 5   item_name       object        
 6   page_name       object        
 7   friend_count    float64       
 8   votes_count     float64       
 9   heart_balance   float64       
 10  question_id     float64       
dtypes: datetime64[ms](1), float64(4), object(6)
memory usage: 960.2+ MB
None


In [38]:
hackle_events_df.isnull().sum()

event_id                 0
event_datetime           0
event_key                0
session_id               0
id                       0
item_name                0
page_name                0
friend_count        752556
votes_count         754554
heart_balance       728643
question_id       10991835
dtype: int64

In [39]:
hackle_events_df.describe()

event_datetime  friend_count   votes_count  heart_balance    question_id
count                    11441319  1.068876e+07  1.068676e+07   1.071268e+07  449484.000000
mean   2023-07-29 01:58:17.236000  5.434357e+01  2.572742e+02   1.626929e+04    2766.385262
min           2023-07-18 00:00:00  0.000000e+00  0.000000e+00   0.000000e+00      99.000000
25%           2023-07-22 21:31:17  3.200000e+01  9.700000e+01   4.340000e+02    1393.000000
50%           2023-07-28 17:08:58  4.900000e+01  2.100000e+02   1.249000e+03    2569.000000
75%           2023-08-04 17:59:10  7.100000e+01  3.620000e+02   3.188000e+03    4459.000000
max           2023-08-10 23:59:59  1.365000e+03  3.017000e+03   8.849998e+08    5133.000000
std                           NaN  3.350798e+01  2.180682e+02   3.317340e+06    1599.967343

In [ ]:
hackle_events_df.describe(include='O')

- question_id는 결측값이 대부분임
- item_name, page_name 은 nan 결측값은 없으나 이상치를 확인해야함

---

In [3]:
hackle_properties_df = pd.read_parquet('hackle/hackle_properties.parquet') 
hackle_properties_df.head()

id                            session_id  user_id language   osname  \
0   1          4OzYh3seq3VKytpSn5pvQkZNQii1             ko-KR      iOS   
1   2          8QXy31PQxbW9qLzq0Y1dhR8Ypm52  1046711    ko-KR      iOS   
2   3  6bcea65d-9f40-46fc-888c-700fe707483f  1545130       ko  Android   
3   4          XVYNT6zfhFWqIg9omwg2AHDjTLx2  1224793       ko  Android   
4   5          XFB2SPiGfjbVhvJ3Q3DBsaT3m2B3  1329450    ko-US      iOS   

  osversion versionname                             device_id  
0      16.0       2.0.0  590E7C79-CBA0-44D8-8BE3-3C9BFABBBC74  
1    16.5.1       2.0.3  D5417226-F71B-4A9E-A180-CD072F2AB279  
2        13       2.0.5  6bcea65d-9f40-46fc-888c-700fe707483f  
3        13       2.0.5  a05c1595-3e05-434b-8684-218b528bd725  
4    16.5.1       2.0.5  EAC6C0B3-7CE8-40EA-8A91-9977C0BA5EF3

In [4]:
hackle_properties_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 525350 entries, 0 to 525349
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype 
---  ------       --------------   ----- 
 0   id           525350 non-null  int64 
 1   session_id   525350 non-null  object
 2   user_id      525350 non-null  object
 3   language     525350 non-null  object
 4   osname       525350 non-null  object
 5   osversion    525350 non-null  object
 6   versionname  525350 non-null  object
 7   device_id    525350 non-null  object
dtypes: int64(1), object(7)
memory usage: 32.1+ MB


In [8]:
hackle_properties_df['user_id'].duplicated().sum()

np.int64(197969)

In [11]:
hackle_properties_df.loc[0, 'user_id']

''

In [12]:
hackle_properties_df[hackle_properties_df['user_id'] == '']

id                            session_id user_id language osname  \
0            1          4OzYh3seq3VKytpSn5pvQkZNQii1            ko-KR    iOS   
41          42          i6toBlZRPDNAnwhvkIWwD16Fcts1            ko-KR    iOS   
126        127          Smot2x4FSsNL1rXLWRGWCpbz2Hw2            ko-KR    iOS   
158        159  3734A994-8BF8-482D-9676-D837DAB4E1E8            ko-KR    iOS   
170        171  6287C8F6-AD13-414B-8320-12C00111E558            ko-KR    iOS   
...        ...                                   ...     ...      ...    ...   
525316  525317  C59904D1-2B1A-4B3A-8318-B00B69C55045            ko-KR    iOS   
525319  525320  1A54B0B9-5FEC-4999-B51B-7349F6B14FB6            en-KR    iOS   
525321  525322  D02C5530-AC9F-46F4-A2A2-D0B3DFB62C01            ko-KR    iOS   
525336  525337          IZUKxswvnIWkNW1JzNE7gBZJGiy2            ko-KR    iOS   
525347  525348  F9EA195C-862F-4572-80A5-C3AE06F83310            ko-KR    iOS   

       osversion versionname                             device_id  
0           16.0       2.0.0  590E7C79-CBA0-44D8-8BE3-3C9BFABBBC74  
41        16.4.1       2.0.5  0C3D3EDA-8F76-43BD-B9AD-7B5DCC3FB8B1  
126       16.5.1       2.0.5  0FB5C35C-EA5B-4D1B-BF14-6874FAF0E582  
158       16.5.1       2.0.5  3734A994-8BF8-482D-9676-D837DAB4E1E8  
170       16.3.1       2.0.5  6287C8F6-AD13-414B-8320-12C00111E558  
...          ...         ...                                   ...  
525316      16.3       2.0.3  C59904D1-2B1A-4B3A-8318-B00B69C55045  
525319    16.5.1       2.0.5  1A54B0B9-5FEC-4999-B51B-7349F6B14FB6  
525321    16.5.1       2.0.3  D02C5530-AC9F-46F4-A2A2-D0B3DFB62C01  
525336      16.1      1.2.10  A3A55D13-20B2-4CAC-ADA5-A70DEB4D4EC4  
525347    16.3.1       2.0.5  F9EA195C-862F-4572-80A5-C3AE06F83310  

[82255 rows x 8 columns]

- user_id의 중복값과 ''인 값들이 많음

---

In [14]:
user_properties_df = pd.read_parquet('hackle/user_properties.parquet') 
user_properties_df.head()

user_id  class gender  grade  school_id
0  1000000      1      M      1       1885
1  1000009     10      F      2       3869
2  1000012     10      F      1       5091
3  1000013      8      F      2       1743
4  1000015      2      F      3       5078

In [15]:
user_properties_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 230819 entries, 0 to 230818
Data columns (total 5 columns):
 #   Column     Non-Null Count   Dtype 
---  ------     --------------   ----- 
 0   user_id    230819 non-null  object
 1   class      230819 non-null  int64 
 2   gender     230819 non-null  object
 3   grade      230819 non-null  int64 
 4   school_id  230819 non-null  int64 
dtypes: int64(3), object(2)
memory usage: 8.8+ MB


In [16]:
user_properties_df['user_id'].duplicated().sum()

np.int64(0)

In [18]:
user_properties_df[user_properties_df['user_id'] == '']

Empty DataFrame
Columns: [user_id, class, gender, grade, school_id]
Index: []

In [19]:
user_properties_df.describe()

class          grade      school_id
count  230819.000000  230819.000000  230819.000000
mean        4.594999       2.002197    3083.383335
std         3.151979       0.762048    1711.671162
min         1.000000       1.000000       1.000000
25%         2.000000       1.000000    1594.000000
50%         4.000000       2.000000    3138.000000
75%         6.000000       3.000000    4640.000000
max        20.000000       3.000000    5964.000000

In [20]:
user_properties_df['school_id'].value_counts().head(10)

school_id
4917    206
5520    206
1471    194
4666    190
461     188
4484    188
465     183
4454    181
1566    180
562     180
Name: count, dtype: int64

In [21]:
user_properties_df['gender'].value_counts()

gender
F    132610
M     98209
Name: count, dtype: int64

- user_id의 중복 / 공백 / nan 결측 없음
- class는 1 ~ 20. 50%가 4인 것으로 보아 큰 수의 class가 적음
- grade는 1~3
- gender는 여자의 비유일 더 많음

---

In [23]:
attendance_df = pd.read_parquet('votes/accounts_attendance.parquet') 
attendance_df.head()

id                               attendance_date_list  user_id
0   1  ["2023-05-27", "2023-05-28", "2023-05-29", "20...  1446852
1   2  ["2023-05-27", "2023-05-29", "2023-05-30", "20...  1359398
2   3  ["2023-05-27", "2023-05-29", "2023-05-30", "20...  1501542
3   4  ["2023-05-27", "2023-05-28", "2023-05-29", "20...  1507767
4   5  ["2023-05-27", "2023-05-28", "2023-05-29", "20...  1287453

In [24]:
attendance_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 349637 entries, 0 to 349636
Data columns (total 3 columns):
 #   Column                Non-Null Count   Dtype 
---  ------                --------------   ----- 
 0   id                    349637 non-null  int64 
 1   attendance_date_list  349637 non-null  object
 2   user_id               349637 non-null  int64 
dtypes: int64(2), object(1)
memory usage: 8.0+ MB


In [25]:
attendance_df['user_id'].duplicated().sum()

np.int64(0)

In [26]:
attendance_df[attendance_df['user_id'] == '']

Empty DataFrame
Columns: [id, attendance_date_list, user_id]
Index: []

- attendance_date_list 는 split 해서 얼마나 자주 방문했는지, 그 빈도는 어떠한지, 주기는 어떠한지 계산할 수 있음
- user_id는 공백 / nan / 중복 없음

---

In [27]:
blockrecord_df = pd.read_parquet('votes/accounts_blockrecord.parquet') 
blockrecord_df.head()

id              reason          created_at  block_user_id  user_id
0   1               그냥... 2023-05-04 23:01:53         867483   878476
1   2         친구 사이가 어색해짐 2023-05-05 01:17:08         867190   867564
2   3  나랑 관련 없는 질문을 자꾸 보냄 2023-05-05 01:50:55         875110   875261
3   6               그냥... 2023-05-05 05:21:52         883696   883511
4   7               그냥... 2023-05-05 06:40:34         871349   870177

In [28]:
blockrecord_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 19482 entries, 0 to 19481
Data columns (total 5 columns):
 #   Column         Non-Null Count  Dtype         
---  ------         --------------  -----         
 0   id             19482 non-null  int64         
 1   reason         19482 non-null  object        
 2   created_at     19482 non-null  datetime64[ns]
 3   block_user_id  19482 non-null  int64         
 4   user_id        19482 non-null  int64         
dtypes: datetime64[ns](1), int64(3), object(1)
memory usage: 761.1+ KB


In [30]:
blockrecord_df['reason'].nunique()

7

In [37]:
blockrecord_df['reason'].value_counts()

reason
모르는 사람임               9640
친구 사이가 어색해짐           5805
사칭 계정                 2022
나랑 관련 없는 질문을 자꾸 보냄    1083
너무 많은 양의 질문을 보냄        919
기타                       7
그냥...                    6
Name: count, dtype: int64

In [31]:
blockrecord_df['block_user_id'].value_counts().head(10)

block_user_id
898020     76
877266     25
897681     25
1380465    25
1395312    24
876207     24
1198628    21
1495281    19
1031842    18
1186363    17
Name: count, dtype: int64

In [33]:
blockrecord_df[blockrecord_df['block_user_id'] == 898020]['reason'].value_counts()

reason
모르는 사람임               25
나랑 관련 없는 질문을 자꾸 보냄    23
너무 많은 양의 질문을 보냄       18
친구 사이가 어색해짐            6
사칭 계정                  4
Name: count, dtype: int64

In [32]:
blockrecord_df['user_id'].value_counts().head(10)

user_id
1526194    171
1415197    163
1519819     70
1542675     66
986459      60
1536035     39
879416      38
1296499     34
1343300     34
1155373     33
Name: count, dtype: int64

In [34]:
blockrecord_df[blockrecord_df['user_id'] == 1526194]['reason'].value_counts()

reason
모르는 사람임    171
Name: count, dtype: int64

In [35]:
blockrecord_df[blockrecord_df['user_id'] == 1415197]['reason'].value_counts()

reason
모르는 사람임        127
친구 사이가 어색해짐     35
사칭 계정            1
Name: count, dtype: int64

In [36]:
blockrecord_df[blockrecord_df['user_id'] == 1519819]['reason'].value_counts()

reason
모르는 사람임        61
친구 사이가 어색해짐     7
사칭 계정           2
Name: count, dtype: int64

- reason 은 7개로 선택 가능해보임
- 특정 유저가 차단을 많이 당하고, 차단을 많이 했던 것으로 확인되어서 이상치 논의 필요
- 차단한 이유 대부분이 모르는 사람이므로, 3위권 안에는 사칭 계정도 많음

---

In [ ]:
# 상품 구매 실패 기록
failpaymenthistory_df = pd.read_parquet('votes/accounts_failpaymenthistory.parquet') 
failpaymenthistory_df.head()

id   productId phone_type          created_at  user_id
0   6   heart.200          A 2023-05-14 05:49:22  1055891
1   7   heart.777          A 2023-05-14 08:17:21  1152151
2   8   heart.777          A 2023-05-14 10:11:46   986200
3   9  heart.1000          A 2023-05-14 11:53:09  1028261
4  10   heart.777          A 2023-05-14 12:30:47  1235730

In [39]:
failpaymenthistory_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 163 entries, 0 to 162
Data columns (total 5 columns):
 #   Column      Non-Null Count  Dtype         
---  ------      --------------  -----         
 0   id          163 non-null    int64         
 1   productId   56 non-null     object        
 2   phone_type  163 non-null    object        
 3   created_at  163 non-null    datetime64[ns]
 4   user_id     163 non-null    int64         
dtypes: datetime64[ns](1), int64(2), object(2)
memory usage: 6.5+ KB


In [40]:
failpaymenthistory_df['productId'].value_counts()

productId
heart.777     49
heart.1000     4
heart.200      3
Name: count, dtype: int64

In [43]:
failpaymenthistory_df[failpaymenthistory_df['user_id'].duplicated()]

id productId phone_type          created_at  user_id
53    61      None          I 2023-05-19 13:38:32  1322662
137  147      None          I 2023-05-31 12:38:35  1211047
162  174      None          I 2023-09-17 09:12:53   927108

In [44]:
failpaymenthistory_df['created_at'].value_counts().head()

created_at
2023-09-17 09:12:53    2
2023-05-29 10:52:40    1
2023-05-25 11:55:06    1
2023-05-25 12:38:08    1
2023-05-26 00:54:31    1
Name: count, dtype: int64

- 결제 실패는 주로 777 에서 발생, 가장 많은건 결측치인 None 타입으로 제거 불가해보임
- 결제 실패를 여러번 겪은 고객은 3명으로 적으며, I 타입의 모바일에서 모두 발생되었음. 

---

In [ ]:
# 학급 테이블
group_df = pd.read_parquet('votes/accounts_group.parquet') 
group_df.head()

id  grade  class_num  school_id
0   1      1          1          1
1   8      3          2          1
2   9      3          3          1
3  10      2          2          1
4  11      2          3          1

In [46]:
group_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 84515 entries, 0 to 84514
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype
---  ------     --------------  -----
 0   id         84515 non-null  int64
 1   grade      84515 non-null  int64
 2   class_num  84515 non-null  int64
 3   school_id  84515 non-null  int64
dtypes: int64(4)
memory usage: 2.6 MB


In [47]:
group_df.describe()

id         grade     class_num     school_id
count  84515.000000  84515.000000  84515.000000  84515.000000
mean   42279.376087      2.113992      5.318606   3079.496468
std    24400.962630      0.754896      4.089901   1675.036093
min        1.000000      1.000000      0.000000      1.000000
25%    21147.500000      2.000000      2.000000   1752.000000
50%    42279.000000      2.000000      4.000000   3023.000000
75%    63411.500000      3.000000      7.000000   4620.000000
max    84546.000000     20.000000     41.000000   5965.000000

In [48]:
group_df['grade'].value_counts()

grade
2     35581
3     29273
1     19659
4         1
20        1
Name: count, dtype: int64

In [49]:
group_df['class_num'].value_counts()

class_num
1     12125
2     11077
3     10276
4      9506
5      8578
6      7555
7      6390
8      5294
9      3799
10     3067
20     2454
11     1600
12      998
13      506
14      311
18      217
15      216
17      214
19      177
16      152
0         2
41        1
Name: count, dtype: int64

- 2, 3 grade가 많고 1은 적으며, 4와 20은 이상치로 봐도 될듯한 수

---

In [50]:
# 가까운 학교를 기록해두기 위한 관계형 테이블
nearbyschool_df = pd.read_parquet('votes/accounts_nearbyschool.parquet') 
nearbyschool_df.head()

id  distance  nearby_school_id  school_id
0  119021  0.004564                 6          7
1  119022  0.010787                13          7
2  119023  0.012928                20          7
3  119024  0.013590                 4          7
4  119025  0.014122                24          7

In [51]:
nearbyschool_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 59500 entries, 0 to 59499
Data columns (total 4 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   id                59500 non-null  int64  
 1   distance          59500 non-null  float64
 2   nearby_school_id  59500 non-null  int64  
 3   school_id         59500 non-null  int64  
dtypes: float64(1), int64(3)
memory usage: 1.8 MB


In [52]:
nearbyschool_df.describe()

id      distance  nearby_school_id     school_id
count   59500.000000  59500.000000      59500.000000  59500.000000
mean   148770.500000      0.055338       2976.695882   2980.524538
std     17176.314845      0.604024       1706.965231   1718.663078
min    119021.000000      0.000000          4.000000      4.000000
25%    133895.750000      0.011484       1518.000000   1493.000000
50%    148770.500000      0.021658       2973.000000   2980.500000
75%    163645.250000      0.064555       4448.000000   4469.000000
max    178520.000000     49.296594       5964.000000   5964.000000

In [53]:
nearbyschool_df['id'].duplicated().sum()

np.int64(0)

- 거리는 0 ~ 0.064rk 75% 차지. max는 49로 이상치 확인 필요

---

In [54]:
# 구매 기록 테이블
paymenthistory_df = pd.read_parquet('votes/accounts_paymenthistory.parquet') 
paymenthistory_df.head()

id  productId phone_type          created_at  user_id
0   6  heart.777          A 2023-05-13 21:28:34  1211127
1   7  heart.777          A 2023-05-13 21:29:39  1151343
2   8  heart.777          A 2023-05-13 21:31:33  1002147
3   9  heart.777          A 2023-05-13 21:31:39  1095040
4  11  heart.777          A 2023-05-13 21:34:32  1164081

In [55]:
paymenthistory_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 95140 entries, 0 to 95139
Data columns (total 5 columns):
 #   Column      Non-Null Count  Dtype         
---  ------      --------------  -----         
 0   id          95140 non-null  int64         
 1   productId   95140 non-null  object        
 2   phone_type  95140 non-null  object        
 3   created_at  95140 non-null  datetime64[ns]
 4   user_id     95140 non-null  int64         
dtypes: datetime64[ns](1), int64(2), object(2)
memory usage: 3.6+ MB


In [56]:
paymenthistory_df['productId'].unique()

array(['heart.777', 'heart.200', 'heart.4000', 'heart.1000'], dtype=object)

In [57]:
paymenthistory_df['productId'].value_counts()

productId
heart.777     57873
heart.1000    19309
heart.200     15822
heart.4000     2136
Name: count, dtype: int64

In [58]:
paymenthistory_df['id'].duplicated().sum()

np.int64(0)

In [59]:
paymenthistory_df['user_id'].duplicated().sum()

np.int64(35948)

In [60]:
paymenthistory_df['user_id'].value_counts().head()

user_id
1527451    60
1246471    51
1141603    35
1204373    34
1142775    30
Name: count, dtype: int64

In [61]:
paymenthistory_df.describe()

id                     created_at       user_id
count  95140.000000                          95140  9.514000e+04
mean   49136.797572  2023-05-27 21:47:32.003741952  1.256912e+06
min        6.000000            2023-05-13 21:28:34  8.330410e+05
25%    24531.750000  2023-05-16 11:48:44.750000128  1.116654e+06
50%    49176.500000     2023-05-21 12:40:45.500000  1.270988e+06
75%    73740.250000  2023-05-27 13:02:52.249999872  1.417660e+06
max    98078.000000            2024-05-08 14:12:45  1.583632e+06
std    28366.854064                            NaN  1.911243e+05

In [62]:
paymenthistory_df.describe(include='O')

productId phone_type
count       95140      95140
unique          4          2
top     heart.777          I
freq        57873      61632

- phone_type 이 I가 가장 많아서 에러도 가장 많이 났던 것으로 유추
- heart.777 구매 비중이 큰 편차로 많음

---

In [63]:
# 포인트 기록 테이블
pointhistory_df = pd.read_parquet('votes/accounts_pointhistory.parquet') 
pointhistory_df.head()

id  delta_point          created_at  user_id  user_question_record_id
0  790629            9 2023-04-28 12:27:49   849436                 771777.0
1  790652            9 2023-04-28 12:28:02   849436                 771800.0
2  790664            5 2023-04-28 12:28:09   849436                 771812.0
3  790680           13 2023-04-28 12:28:16   849436                 771828.0
4  790703            5 2023-04-28 12:28:26   849436                 771851.0

In [64]:
pointhistory_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2338918 entries, 0 to 2338917
Data columns (total 5 columns):
 #   Column                   Dtype         
---  ------                   -----         
 0   id                       int64         
 1   delta_point              int64         
 2   created_at               datetime64[ns]
 3   user_id                  int64         
 4   user_question_record_id  float64       
dtypes: datetime64[ns](1), float64(1), int64(3)
memory usage: 89.2 MB


In [65]:
pointhistory_df.describe()

id   delta_point                     created_at  \
count  2.338918e+06  2.338918e+06                        2338918   
mean   1.265364e+08  8.610862e-01  2023-05-18 03:06:43.825683968   
min    7.906290e+05 -1.000000e+03            2023-04-28 12:27:49   
25%    4.112505e+07  7.000000e+00            2023-05-11 04:37:02   
50%    1.123971e+08  1.000000e+01            2023-05-16 07:29:11   
75%    1.982051e+08  1.300000e+01  2023-05-22 13:47:48.750000128   
max    3.406724e+08  1.000000e+03            2024-05-08 01:36:18   
std    9.610992e+07  5.885849e+01                            NaN   

            user_id  user_question_record_id  
count  2.338918e+06             2.335926e+06  
mean   1.111554e+06             6.141803e+07  
min    8.380230e+05             7.717770e+05  
25%    8.861890e+05             2.088561e+07  
50%    1.126012e+06             5.540614e+07  
75%    1.260221e+06             9.555855e+07  
max    1.583358e+06             1.616665e+08  
std    2.024429e+05             4.531970e+07

- delta_point에 마이너스도 있음

---

In [66]:
# 학교 테이블
school_df = pd.read_parquet('votes/accounts_school.parquet') 
school_df.head()

id   address  student_count school_type
0   4  충청북도 충주시            239           H
1   5  충청북도 충주시            160           M
2   6  충청북도 충주시            200           H
3   7  충청북도 충주시            114           H
4   8  충청북도 충주시            139           M

In [67]:
school_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5951 entries, 0 to 5950
Data columns (total 4 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   id             5951 non-null   int64 
 1   address        5951 non-null   object
 2   student_count  5951 non-null   int64 
 3   school_type    5951 non-null   object
dtypes: int64(2), object(2)
memory usage: 186.1+ KB


In [68]:
school_df['address'].nunique()

278

In [69]:
school_df['school_type'].value_counts()

school_type
M    3290
H    2661
Name: count, dtype: int64

In [70]:
school_df.describe()

id  student_count
count  5951.000000    5951.000000
mean   2981.026046     113.772979
std    1719.084020     102.863428
min       4.000000       0.000000
25%    1493.500000      16.000000
50%    2981.000000      97.000000
75%    4469.500000     183.000000
max    5965.000000     578.000000

In [71]:
school_df.describe(include='O')

address school_type
count      5951        5951
unique      278           2
top     경기도 화성시           M
freq         78        3290

- 경기도 화성시가 제일 큰 비중
- 중/고등학교의 비율이 크게 차이나진 않음. 중학교가 더 많음
- 학생수의 편차가 커서 이상치 확인 필요

---

In [72]:
# 유저 신고 기록 테이블
timelinereport_df = pd.read_parquet('votes/accounts_timelinereport.parquet') 
timelinereport_df.head()

id             reason          created_at  reported_user_id  user_id  \
0  28            타인을 사칭함 2023-05-06 04:44:57            874587   885082   
1  37  친구를 비하하거나 조롱하는 어투 2023-05-06 05:41:19            881048   881298   
2  46  친구를 비하하거나 조롱하는 어투 2023-05-06 06:26:59            887882   881945   
3  64           허위 사실 언급 2023-05-06 07:07:46            888610   893684   
4  65           허위 사실 언급 2023-05-06 07:07:52            888610   893684   

   user_question_record_id  
0                  3920588  
1                  4018679  
2                  4120914  
3                  4143049  
4                  4143049

In [73]:
timelinereport_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 208 entries, 0 to 207
Data columns (total 6 columns):
 #   Column                   Non-Null Count  Dtype         
---  ------                   --------------  -----         
 0   id                       208 non-null    int64         
 1   reason                   208 non-null    object        
 2   created_at               208 non-null    datetime64[ns]
 3   reported_user_id         208 non-null    int64         
 4   user_id                  208 non-null    int64         
 5   user_question_record_id  208 non-null    int64         
dtypes: datetime64[ns](1), int64(4), object(1)
memory usage: 9.9+ KB


In [75]:
timelinereport_df['reason'].value_counts()

reason
허위 사실 언급             80
친구를 비하하거나 조롱하는 어투    64
선정적이거나 폭력적인 내용       39
타인을 사칭함              15
광고                   10
Name: count, dtype: int64

- 허위 사실 언급 신고가 가장 많음.

---

In [76]:
# 유저의 네트워크 정보 테이블 (연락처, 초대)
contacts_df = pd.read_parquet('votes/accounts_user_contacts.parquet') 
contacts_df.head()

id  contacts_count invite_user_id_list  user_id
0    259              30                  []  1167696
1   1756              79                  []   863169
2  13742              21            [854615]   857205
3  13754              29                  []   851431
4  13756              28            [849318]   855476

In [77]:
contacts_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5063 entries, 0 to 5062
Data columns (total 4 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   id                   5063 non-null   int64 
 1   contacts_count       5063 non-null   int64 
 2   invite_user_id_list  5063 non-null   object
 3   user_id              5063 non-null   int64 
dtypes: int64(3), object(1)
memory usage: 158.3+ KB


In [78]:
contacts_df['user_id'].duplicated().sum()

np.int64(0)

In [79]:
contacts_df['invite_user_id_list'].nunique()

928

---

In [80]:
# 유저 테이블
user_df = pd.read_parquet('votes/accounts_user.parquet') 
user_df.head()

id  is_superuser  is_staff gender  point  \
0  831956             1         1   None    600   
1  831962             0         0      F   2248   
2  832151             0         0      M   1519   
3  832340             0         0      F     57   
4  832520             0         0      M   1039   

                                      friend_id_list  is_push_on  \
0  [1292473, 913158, 1488461, 1064695, 1043565, 1...           0   
1  [833025, 832642, 982531, 879496, 838541, 83752...           1   
2  [838785, 982531, 882567, 879496, 838541, 83649...           0   
3  [841345, 982531, 838785, 963714, 882567, 83252...           1   
4  [874050, 849763, 874212, 844297, 838541, 84004...           0   

                  created_at block_user_id_list hide_user_id_list ban_status  \
0 2023-03-29 03:44:14.047130                 []                []          N   
1 2023-03-29 05:18:56.162368                 []                []          N   
2 2023-03-29 12:56:34.989468                 []                []          N   
3 2023-03-29 12:56:35.020790                 []                []          N   
4 2023-03-29 12:56:35.049311                 []                []          N   

   report_count  alarm_count  pending_chat  pending_votes  group_id  
0             0            0             0              0       NaN  
1           253        40878          5499            110      12.0  
2             0           37             0             47       1.0  
3             0           19             0             21       1.0  
4             0           29             0             15      12.0

In [81]:
user_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 677085 entries, 0 to 677084
Data columns (total 16 columns):
 #   Column              Non-Null Count   Dtype         
---  ------              --------------   -----         
 0   id                  677085 non-null  int64         
 1   is_superuser        677085 non-null  int64         
 2   is_staff            677085 non-null  int64         
 3   gender              677083 non-null  object        
 4   point               677085 non-null  int64         
 5   friend_id_list      677085 non-null  object        
 6   is_push_on          677085 non-null  int64         
 7   created_at          677085 non-null  datetime64[ns]
 8   block_user_id_list  677085 non-null  object        
 9   hide_user_id_list   677085 non-null  object        
 10  ban_status          677085 non-null  object        
 11  report_count        677085 non-null  int64         
 12  alarm_count         677085 non-null  int64         
 13  pending_chat        677085 no

In [86]:
user_df['gender'].value_counts()

gender
F    396665
M    280418
Name: count, dtype: int64

In [87]:
user_df['is_staff'].value_counts()

is_staff
0    677082
1         3
Name: count, dtype: int64

In [88]:
user_df['is_push_on'].value_counts()

is_push_on
1    570849
0    106236
Name: count, dtype: int64

In [89]:
user_df['is_superuser'].value_counts()

is_superuser
0    677083
1         2
Name: count, dtype: int64

- gender, group_id 에만 3개 이하의 nan 결측치
- 대부분이 superuser
- 여성 유저가 더 많음
- 알람 동의가 더 많음
- 친구 리스트, 차단 리스트 확인 가능

---

In [90]:
# 투표 기록 테이블
userquestionrecord_df = pd.read_parquet('votes/accounts_userquestionrecord.parquet') 
userquestionrecord_df.head()

id status          created_at  chosen_user_id  question_id  user_id  \
0  771777      C 2023-04-28 12:27:49          849469          252   849436   
1  771800      C 2023-04-28 12:28:02          849446          244   849436   
2  771812      C 2023-04-28 12:28:09          849454          183   849436   
3  771828      C 2023-04-28 12:28:16          847375          101   849436   
4  771851      C 2023-04-28 12:28:26          849477          209   849436   

   question_piece_id  has_read answer_status   answer_updated_at  \
0             998458         0             N 2023-04-28 12:27:49   
1             998459         0             N 2023-04-28 12:28:02   
2             998460         1             N 2023-04-28 12:28:09   
3             998461         0             N 2023-04-28 12:28:16   
4             998462         1             N 2023-04-28 12:28:26   

   report_count  opened_times  
0             0             0  
1             0             0  
2             0             0  
3             0             0  
4             0             0

In [91]:
userquestionrecord_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1217558 entries, 0 to 1217557
Data columns (total 12 columns):
 #   Column             Non-Null Count    Dtype         
---  ------             --------------    -----         
 0   id                 1217558 non-null  int64         
 1   status             1217558 non-null  object        
 2   created_at         1217558 non-null  datetime64[ns]
 3   chosen_user_id     1217558 non-null  int64         
 4   question_id        1217558 non-null  int64         
 5   user_id            1217558 non-null  int64         
 6   question_piece_id  1217558 non-null  int64         
 7   has_read           1217558 non-null  int64         
 8   answer_status      1217558 non-null  object        
 9   answer_updated_at  1217558 non-null  datetime64[ns]
 10  report_count       1217558 non-null  int64         
 11  opened_times       1217558 non-null  int64         
dtypes: datetime64[ns](2), int64(8), object(2)
memory usage: 111.5+ MB


In [92]:
userquestionrecord_df['question_piece_id'].duplicated().sum()

np.int64(0)

In [93]:
userquestionrecord_df['question_id'].duplicated().sum()

np.int64(1212753)

In [94]:
userquestionrecord_df['status'].value_counts()

status
C    1156322
I      60578
B        658
Name: count, dtype: int64

In [95]:
userquestionrecord_df.describe()

id                     created_at  chosen_user_id  \
count  1.217558e+06                        1217558    1.217558e+06   
mean   5.957298e+07  2023-05-17 12:51:25.947102208    1.092604e+06   
min    7.717770e+05            2023-04-28 12:27:49    8.331120e+05   
25%    1.553965e+07            2023-05-10 01:42:11    8.836920e+05   
50%    5.302680e+07     2023-05-15 15:43:01.500000    1.091749e+06   
75%    9.480960e+07  2023-05-22 11:01:03.750000128    1.235698e+06   
max    1.616665e+08            2024-05-08 01:36:18    1.579422e+06   
std    4.608193e+07                            NaN    2.027585e+05   

        question_id       user_id  question_piece_id      has_read  \
count  1.217558e+06  1.217558e+06       1.217558e+06  1.217558e+06   
mean   6.843895e+02  1.105790e+06       7.413211e+07  5.551530e-01   
min    9.900000e+01  8.380230e+05       9.984580e+05  0.000000e+00   
25%    2.750000e+02  8.846190e+05       1.854142e+07  0.000000e+00   
50%    4.690000e+02  1.117537e+06       6.616843e+07  1.000000e+00   
75%    9.420000e+02  1.259186e+06       1.176735e+08  1.000000e+00   
max    5.133000e+03  1.583358e+06       2.083515e+08  1.000000e+00   
std    6.250399e+02  2.062704e+05       5.757237e+07  4.969490e-01   

                   answer_updated_at  report_count  opened_times  
count                        1217558  1.217558e+06  1.217558e+06  
mean   2023-05-17 13:44:38.920359680  1.765830e-04  6.332511e-02  
min              2023-04-28 12:27:49  0.000000e+00  0.000000e+00  
25%    2023-05-10 02:38:44.750000128  0.000000e+00  0.000000e+00  
50%       2023-05-15 16:29:08.500000  0.000000e+00  0.000000e+00  
75%    2023-05-22 11:41:54.750000128  0.000000e+00  0.000000e+00  
max              2024-05-08 01:36:18  1.400000e+01  3.000000e+00  
std                              NaN  2.040530e-02  3.011468e-01

- question_id 는 중복 존재. question_piece_id는 중복 없음.
- C 상태가 가장 많음

---

In [96]:
# 탈퇴 기록 테이블
userwithdraw_df = pd.read_parquet('votes/accounts_userwithdraw.parquet') 
userwithdraw_df.head()

id        reason          created_at
0  42   재밌는 질문이 없어서 2023-03-29 13:22:12
1  43  함께 할 친구가 없어서 2023-03-29 13:24:59
2  44  함께 할 친구가 없어서 2023-03-29 13:53:32
3  45  함께 할 친구가 없어서 2023-03-29 13:54:03
4  46   재밌는 질문이 없어서 2023-03-30 00:31:19

In [97]:
userwithdraw_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 70764 entries, 0 to 70763
Data columns (total 3 columns):
 #   Column      Non-Null Count  Dtype         
---  ------      --------------  -----         
 0   id          70764 non-null  int64         
 1   reason      70764 non-null  object        
 2   created_at  70764 non-null  datetime64[ns]
dtypes: datetime64[ns](1), int64(1), object(1)
memory usage: 1.6+ MB


In [98]:
userwithdraw_df['reason'].value_counts()

reason
기타 이유           40301
함께 할 친구가 없어서    14450
재밌는 질문이 없어서     13133
버그가 너무 많아서       2031
구독료가 너무 비싸서       730
admin              61
test               53
기타                  5
Name: count, dtype: int64

---

In [99]:
# 포인트 이벤트 참여 테이블
event_receipts_df = pd.read_parquet('votes/event_receipts.parquet') 
event_receipts_df.head()

id          created_at  event_id  user_id  plus_point
0   2 2023-06-22 09:25:16         1  1193618         500
1   3 2023-06-22 09:38:53         1   928351         500
2   4 2023-06-22 10:32:15         1   904872         500
3   5 2023-06-22 13:03:06         1   974697         500
4   6 2023-06-22 13:40:38         1  1168260         500

In [100]:
event_receipts_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 309 entries, 0 to 308
Data columns (total 5 columns):
 #   Column      Non-Null Count  Dtype         
---  ------      --------------  -----         
 0   id          309 non-null    int64         
 1   created_at  309 non-null    datetime64[ns]
 2   event_id    309 non-null    int64         
 3   user_id     309 non-null    int64         
 4   plus_point  309 non-null    int64         
dtypes: datetime64[ns](1), int64(4)
memory usage: 12.2 KB


In [101]:
event_receipts_df['event_id'].value_counts()

event_id
1    252
2     37
3     20
Name: count, dtype: int64

In [103]:
event_receipts_df[event_receipts_df['user_id'].duplicated()]

id          created_at  event_id  user_id  plus_point
76  81 2023-07-10 16:57:04         1  1577954         500

In [104]:
event_receipts_df.describe()

id                     created_at    event_id       user_id  \
count  309.000000                            309  309.000000  3.090000e+02   
mean   160.964401  2023-07-24 01:53:59.954692608    1.249191  1.179715e+06   
min      2.000000            2023-06-22 09:25:16    1.000000  8.378360e+05   
25%     82.000000            2023-07-10 16:58:20    1.000000  9.998160e+05   
50%    161.000000            2023-07-20 12:24:28    1.000000  1.139491e+06   
75%    239.000000            2023-07-29 09:38:22    1.000000  1.325911e+06   
max    321.000000            2023-11-21 12:03:02    3.000000  1.579064e+06   
std     92.065029                            NaN    0.563536  2.171981e+05   

        plus_point  
count   309.000000  
mean    559.870550  
min     500.000000  
25%     500.000000  
50%     500.000000  
75%     500.000000  
max    1000.000000  
std     162.592579

In [108]:
event_receipts_df['plus_point'].value_counts()

plus_point
500     272
1000     37
Name: count, dtype: int64

In [110]:
event_receipts_df[event_receipts_df['plus_point'] == 1000]['event_id'].value_counts()

event_id
2    37
Name: count, dtype: int64

- 1번 이벤트 참여가 제일 높음 (이벤트는 1~3)
- 포인트는 500 ~ 1000 이고 대부분 500
    - 1000포인트는 2번 이벤트 뿐

---

In [107]:
# 포인트 이벤트 테이블
events_df = pd.read_parquet('votes/events.parquet') 
events_df

id           title  plus_point event_type  is_expired          created_at
0   1   코드잇 은행 가입 이벤트         500       FCFS           1 2023-06-20 11:56:38
1   2  코드잇 멤버십 가입 이벤트        1000       FCFS           1 2023-08-08 07:43:45
2   3   예고 영상 기대평 이벤트         500       FCFS           1 2023-09-24 17:05:59

In [106]:
events_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3 entries, 0 to 2
Data columns (total 6 columns):
 #   Column      Non-Null Count  Dtype         
---  ------      --------------  -----         
 0   id          3 non-null      int64         
 1   title       3 non-null      object        
 2   plus_point  3 non-null      int64         
 3   event_type  3 non-null      object        
 4   is_expired  3 non-null      int64         
 5   created_at  3 non-null      datetime64[ns]
dtypes: datetime64[ns](1), int64(3), object(2)
memory usage: 272.0+ bytes


- 참여율 좋았던 1번 이벤트는 코드잇 은행 가입 이벤트
- 이벤트 중복 참여는 안되는 것으로 확인됨

---

In [111]:
# 질문 내용 테이블
polls_question_df = pd.read_parquet('votes/polls_question.parquet') 
polls_question_df.head()

id                 question_text          created_at
0   99            가장 신비한 매력이 있는 사람은? 2023-03-31 15:22:53
1  100  "이 사람으로 한 번 살아보고 싶다" 하는 사람은? 2023-03-31 15:22:53
2  101                     미래의 틱톡커는? 2023-03-31 15:22:54
3  102               여기서 제일 특이한 친구는? 2023-03-31 15:22:54
4  103               가장 지켜주고 싶은 사람은? 2023-03-31 15:22:55

In [112]:
polls_question_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5025 entries, 0 to 5024
Data columns (total 3 columns):
 #   Column         Non-Null Count  Dtype         
---  ------         --------------  -----         
 0   id             5025 non-null   int64         
 1   question_text  5025 non-null   object        
 2   created_at     5025 non-null   datetime64[ns]
dtypes: datetime64[ns](1), int64(1), object(1)
memory usage: 117.9+ KB


In [113]:
polls_question_df['question_text'].nunique()

3903

In [114]:
polls_question_df['id'].nunique()

5025

---

In [115]:
# 질문 테이블
polls_questionpiece_df = pd.read_parquet('votes/polls_questionpiece.parquet') 
polls_questionpiece_df.head()

id  is_voted          created_at  question_id  is_skipped
0  998458         1 2023-04-28 12:27:22          252           0
1  998459         1 2023-04-28 12:27:22          244           0
2  998460         1 2023-04-28 12:27:22          183           0
3  998461         1 2023-04-28 12:27:22          101           0
4  998462         1 2023-04-28 12:27:22          209           0

In [116]:
polls_questionpiece_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1265476 entries, 0 to 1265475
Data columns (total 5 columns):
 #   Column       Non-Null Count    Dtype         
---  ------       --------------    -----         
 0   id           1265476 non-null  int64         
 1   is_voted     1265476 non-null  int64         
 2   created_at   1265476 non-null  datetime64[ns]
 3   question_id  1265476 non-null  int64         
 4   is_skipped   1265476 non-null  int64         
dtypes: datetime64[ns](1), int64(4)
memory usage: 48.3 MB


In [117]:
polls_questionpiece_df['question_id'].duplicated().sum()

np.int64(1260532)

In [118]:
polls_questionpiece_df.describe()

id      is_voted                     created_at  \
count  1.265476e+06  1.265476e+06                        1265476   
mean   7.658001e+07  9.630266e-01  2023-05-17 13:39:58.427317504   
min    9.984580e+05  0.000000e+00            2023-04-28 12:27:22   
25%    2.029260e+07  1.000000e+00            2023-05-09 21:50:54   
50%    6.846533e+07  1.000000e+00     2023-05-15 14:18:31.500000   
75%    1.212662e+08  1.000000e+00            2023-05-22 09:56:44   
max    2.083852e+08  1.000000e+00            2024-05-07 11:32:30   
std    5.912924e+07  1.886967e-01                            NaN   

        question_id    is_skipped  
count  1.265476e+06  1.265476e+06  
mean   7.133598e+02  8.905740e-04  
min    9.900000e+01  0.000000e+00  
25%    2.780000e+02  0.000000e+00  
50%    4.770000e+02  0.000000e+00  
75%    9.820000e+02  0.000000e+00  
max    5.133000e+03  1.000000e+00  
std    6.706815e+02  2.982921e-02

In [119]:
polls_questionpiece_df['is_skipped'].value_counts()

is_skipped
0    1264349
1       1127
Name: count, dtype: int64

In [120]:
polls_questionpiece_df['is_voted'].value_counts()

is_voted
1    1218687
0      46789
Name: count, dtype: int64

- 질문 id는 중복이 많음
- 스킵 없음이 비중이 큼
- 투표 비중이 큼

---

In [121]:
# 질문에 대한 신고 기록 테이블
polls_questionreport_df = pd.read_parquet('votes/polls_questionreport.parquet') 
polls_questionreport_df.head()

id                reason          created_at  question_id  user_id
0   1           이 질문은 재미없어요 2023-04-19 06:20:35          250   837556
1   2           이 질문은 재미없어요 2023-04-19 06:58:09          113   837672
2   3       불쾌한 내용이 포함되어 있음 2023-04-19 06:58:17          113   837672
3   4  어떻게 이런 생각을? 이 질문 최고! 2023-04-19 08:12:42          119   837922
4   5  어떻게 이런 생각을? 이 질문 최고! 2023-04-19 08:12:50          119   837922

In [122]:
polls_questionreport_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 51424 entries, 0 to 51423
Data columns (total 5 columns):
 #   Column       Non-Null Count  Dtype         
---  ------       --------------  -----         
 0   id           51424 non-null  int64         
 1   reason       51424 non-null  object        
 2   created_at   51424 non-null  datetime64[ns]
 3   question_id  51424 non-null  int64         
 4   user_id      51424 non-null  int64         
dtypes: datetime64[ns](1), int64(3), object(1)
memory usage: 2.0+ MB


In [127]:
polls_questionreport_df[polls_questionreport_df['question_id'].isin(polls_questionreport_df['question_id'].value_counts().head())]

id                reason          created_at  question_id  user_id
1983    2136  어떻게 이런 생각을? 이 질문 최고! 2023-05-02 12:45:23          533   863947
2216    2382           이 질문은 재미없어요 2023-05-03 03:12:14          533   859534
2649    2839  어떻게 이런 생각을? 이 질문 최고! 2023-05-04 12:23:43          493   875572
2973    3194  어떻게 이런 생각을? 이 질문 최고! 2023-05-05 13:41:25          493   884403
2983    3204  어떻게 이런 생각을? 이 질문 최고! 2023-05-05 13:55:29          493   886404
...      ...                   ...                 ...          ...      ...
46891  51013                 그냥 싫어 2023-06-03 05:00:01          988   949231
47321  51493             불쾌한 질문 내용 2023-06-04 04:41:34          533   917248
49623  53922                 그냥 싫어 2023-06-24 08:06:09          988  1441146
51230  55574                 그냥 싫어 2023-10-31 13:11:33          493   928351
51291  55635                 그냥 싫어 2023-12-18 11:14:02          493   942148

[191 rows x 5 columns]

In [128]:
polls_questionreport_df['reason'].value_counts()

reason
그냥 싫어                   28446
나랑 맞지 않는 질문인 것 같음        9541
불쾌한 질문 내용                5386
자꾸 같은 내용의 질문 반복          3202
어떻게 이런 생각을? 이 질문 최고!     1821
한 친구가 질문을 반복적으로 보냄       1701
기타                        480
이 질문은 재미없어요               471
불쾌한 내용이 포함되어 있음           250
오타가 있음                     68
선정적이거나 자극적인 질문             58
Name: count, dtype: int64

- 신고에 어울리지 않는 칭찬 신고도 있음.

---

In [129]:
# 질문 세트 테이블
polls_questionset_df = pd.read_parquet('votes/polls_questionset.parquet') 
polls_questionset_df.head()

id                             question_piece_id_list  \
0  99817  [998458, 998459, 998460, 998461, 998462, 99846...   
1  99830  [998588, 998589, 998590, 998591, 998592, 99859...   
2  99840  [998689, 998691, 998693, 998695, 998697, 99869...   
3  99841  [998688, 998690, 998692, 998694, 998696, 99869...   
4  99848  [998768, 998769, 998770, 998771, 998772, 99877...   

         opening_time status          created_at  user_id  
0 2023-04-28 12:27:22      F 2023-04-28 12:27:23   849436  
1 2023-04-28 12:28:07      F 2023-04-28 12:28:07   849438  
2 2023-04-28 12:28:38      F 2023-04-28 12:28:38   847375  
3 2023-04-28 12:28:38      F 2023-04-28 12:28:38   849446  
4 2023-04-28 12:28:57      F 2023-04-28 12:28:57   849477

In [130]:
polls_questionset_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 158384 entries, 0 to 158383
Data columns (total 6 columns):
 #   Column                  Non-Null Count   Dtype         
---  ------                  --------------   -----         
 0   id                      158384 non-null  int64         
 1   question_piece_id_list  158384 non-null  object        
 2   opening_time            158384 non-null  datetime64[ns]
 3   status                  158384 non-null  object        
 4   created_at              158384 non-null  datetime64[ns]
 5   user_id                 158384 non-null  int64         
dtypes: datetime64[ns](2), int64(2), object(2)
memory usage: 7.3+ MB


In [131]:
polls_questionset_df['user_id'].duplicated().sum()

np.int64(153412)

In [132]:
polls_questionset_df['status'].value_counts()

status
F    153411
O      4407
C       566
Name: count, dtype: int64

---

In [134]:
# 질문에 등장하는 유저 테이블
polls_usercandidate_df = pd.read_parquet('votes/polls_usercandidate.parquet') 
polls_usercandidate_df.head()

id          created_at  question_piece_id  user_id
0  3088872 2023-04-28 12:27:49             998458   849444
1  3088873 2023-04-28 12:27:49             998458   849454
2  3088874 2023-04-28 12:27:49             998458   849460
3  3088875 2023-04-28 12:27:49             998458   849469
4  3088964 2023-04-28 12:28:02             998459   849446

In [135]:
polls_usercandidate_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4769609 entries, 0 to 4769608
Data columns (total 4 columns):
 #   Column             Dtype         
---  ------             -----         
 0   id                 int64         
 1   created_at         datetime64[ns]
 2   question_piece_id  int64         
 3   user_id            int64         
dtypes: datetime64[ns](1), int64(3)
memory usage: 145.6 MB


In [136]:
polls_usercandidate_df.describe()

id                     created_at  question_piece_id  \
count  4.769609e+06                        4769609       4.769609e+06   
mean   2.388684e+08  2023-05-17 13:57:52.437731072       7.430248e+07   
min    3.088872e+06            2023-04-28 12:27:49       9.984580e+05   
25%    6.270339e+07            2023-05-10 02:49:19       1.891695e+07   
50%    2.126700e+08            2023-05-15 16:15:12       6.639655e+07   
75%    3.800215e+08            2023-05-22 11:33:33       1.178876e+08   
max    6.466726e+08            2024-05-08 01:36:18       2.083515e+08   
std    1.844893e+08                            NaN       5.761959e+07   

            user_id  
count  4.769609e+06  
mean   1.090241e+06  
min    8.331120e+05  
25%    8.835900e+05  
50%    1.087460e+06  
75%    1.233969e+06  
max    1.580629e+06  
std    2.018190e+05

---